contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,821 @@
+        // the StatefulSet when Type is PartitionedStatefulSetStrategyType. This 
+        // value must be set when Type is PartitionedStatefulSetStrategyType, 
+        // and it must be nil otherwise.
+        Partition *PartitionedStatefulSet
+
+// StatefulSetUpdateStrategyType is a string enumeration type that enumerates 
+// all possible update strategies for the StatefulSet controller.
+type StatefulSetUpdateStrategyType string
+
+const (
+    // PartitionedStatefulSetStrategyType indicates that updates will only be 
+    // applied to a partition of the StatefulSet. This is useful for canaries 
+    // and phased roll outs.
+    PartitionedStatefulSetStrategyType StatefulSetUpdateStrategyType = "Partitioned"
+    // RollingUpdateStatefulSetStrategyType indicates that update will be 
+    // applied to all Pods in the StatefulSet with respect to the StatefulSet 
+    // ordering constraints.
+    RollingUpdateStatefulSetStrategyType = "RollingUpdate"
+    // OnDeleteStatefulSetStrategyType triggers the legacy behavior. Version 
+    // tracking and ordered rolling restarts are disabled. Pods are recreated 
+    // from the StatefulSetSpec when they are manually deleted.
+    OnDeleteStatefulSetStrategyType = "OnDelete"
+)
+
+// PartitionedStatefulSet contains the parameters used with the 
+// PartitionedStatefulSetStrategyType.
+type PartitionedStatefulSet struct {
+    // Ordinal indicates the ordinal at which the StatefulSet should be 
+    // partitioned.
+    Ordinal int32
+}
+
+type StatefulSetSpec struct {
+    // Replicas, Selector, Template, VolumeClaimsTemplate, and ServiceName 
+    // omitted for brevity.
+
+	// UpdateStrategy indicates the StatefulSetUpdateStrategy that will be 
+	// employed to update Pods in the StatefulSet when a revision is made to 
+	// Template or VolumeClaimsTemplate.
+	UpdateStrategy StatefulSetUpdateStrategy `json:"updateStrategy,omitempty`
+
+	// RevisionHistoryLimit is the maximum number of PodTemplates that will 
+	// be maintained in the StatefulSet's revision history. The revision history
+	// consists of all revisions not represented by a currently applied 
+	// StatefulSetSpec version. The default value is 2.
+	RevisionHistoryLimit *int32 `json:revisionHistoryLimit,omitempty`
+}
+```
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are omitted for brevity.
+
+ 	// CurrentRevision, if not empty, indicates the version of PodSpecTemplate, 
+ 	// VolumeClaimsTemplate tuple used to generate Pods in the sequence
+ 	// [0,CurrentReplicas).
+ 	CurrentRevision string `json:"currentRevision,omitempty"`
+ 	
+ 	// UpdateRevision, if not empty, indicates the version of PodSpecTemplate, 
+    // VolumeClaimsTemplate tuple used to generate Pods in the sequence
+    // [Replicas-UpdatedReplicas,Replicas)
+ 	UpdateRevision string `json:"updateRevision,omitempty"`
```

Comment:
@kow3ns @kargakis one thing not clear to me is if 0---CurrentReplicas are running with Current Revision and Replicas-UpdatedReplicas--- Replicas are running with UpdatedRevision, then what about the pods from CurrentReplicas ----- Replicas-UpdatedReplicas, what version are they running with ? May be for both partitioned and rolling update case, show an in progress update and the corresponding values so that its easy to reason about which pods are which version